nucypher/network/server.py

```diff
@@ -334,73 +336,92 @@ def reencrypt_via_rest(id_as_hex):
         else:
-            return Response("Can't save a TreasureMap with this ID from you.", status=409)
+            map_identifier = received_treasure_map.public_id()
+            expiration_date = MayaDT.from_datetime(datetime.utcnow() + timedelta(days=7))
 
-        if do_store and not this_node.federated_only:
-            alice_checksum_address = this_node.policy_agent.contract.functions.getPolicyOwner(
-                treasure_map._hrac[:16]).call()
-            do_store = treasure_map.verify_blockchain_signature(checksum_address=alice_checksum_address)
+        # Step 2: Check if we already have the treasure map.
+        try:
+            with datastore.describe(TreasureMap, map_identifier) as stored_treasure_map:
+                if _MapClass.from_bytes(stored_treasure_map.treasure_map) == received_treasure_map:
+                    return Response("Already have this map.", status=303)
+        except RecordNotFound:
+            # This appears to be a new treasure map that we don't have!
+            pass
 
-        if do_store:
-            log.info("{} storing TreasureMap {}".format(this_node, treasure_map_id))
-            this_node.treasure_maps[treasure_map_index] = treasure_map
-            return Response(bytes(treasure_map), status=202)
-        else:
-            log.info("Bad TreasureMap ID; not storing {}".format(treasure_map_id))
-            return Response("This TreasureMap doesn't match a paid Policy.", status=402)
+        # Step 3: If the node is decentralized, we check that the received
+        # treasure map is valid pursuant to an active policy.
+        # We also set the expiration from the data on the blockchain here.
+        if not this_node.federated_only:
+            policy_data, alice_checksum_address = this_node.policy_agent.fetch_policy(
+                    received_treasure_map._hrac,
+                    with_owner=True)
+            # If the Policy doesn't exist, the policy_data is all zeros.
+            if not policy_data[5]:
+                log.info(f"TreasureMap is for non-existent Policy; not storing {map_identifier}")
+                return Response("The Policy for this TreasureMap doesn't exist.", status=409)
+
+            # Check that this treasure map is from Alice per the Policy.
+            if not received_treasure_map.verify_blockchain_signature(checksum_address=alice_checksum_address):
+                log.info(f"Bad TreasureMap ID; not storing {map_identifier}")
+                return Response("This TreasureMap doesn't match a paid Policy.", status=402)
+
+            # Check that this treasure map is valid for the Policy datetime and that it's not disabled.
+            if policy_data[0] or datetime.utcnow() >= datetime.utcfromtimestamp(policy_data[5]):
+                log.info(f"Received TreasureMap for an expired/disabled policy; not storing {map_identifier}")
+                return Response("This TreasureMap is for an expired/disabled policy.", status=402)
```

Comment:
No, because the node will happily accept a treasure map if there's a payment/policy for it. Sorta an inbetween, but I think a 402 is much more truthy.